### PR TITLE
DAOS-3417 control: allow config path option to be relative

### PR DESCRIPTION
**Pull request by tanabarr**
Allow relative (to user's working directory) config file path to be
specified as cli option to daos_{shell,server,agent}.

Improve consistency of setting config path across client and server by
introducing SetPath() methods and move duplicated logic to common
package with unit tests.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

**Comment by tanabarr** [comment]
replaces https://github.com/daos-stack/daos/pull/1228

**Comment by tanabarr** [comment]
supplying relative path to server:
```
[root@wolf-71 daos_m]# daos_server start -o utils/config/examples/daos_server_sockets.yml -i --debug                    DEBUG 10:40:21.451332 main.go:110: DAOS config loaded from /home/tanabarr/projects/daos_m/utils/config/examples/daos_server_sockets.yml
...
``
supplying relative path to dmg:
```
tanabarr@wolf-71 daos_m]$ dmg -i storage scan -o utils/config/daos.yml --debug
DEBUG 10:50:00.148520 main.go:130: debug output enabled
DEBUG 10:50:00.148893 config.go:119: DAOS Client config read from /home/tanabarr/projects/daos_m/utils/config/daos.yml
localhost:10001: connected
Hosts           SCM Total               NVMe Total
-----           ---------               ----------
localhost       5.79TB (2 namespaces)   3.03TB (3 controllers)
```
supplying relative path to agent:
```
tanabarr@wolf-71 daos_m]$ dmg -i storage scan -o utils/config/daos.yml --debug
DEBUG 10:50:00.148520 main.go:130: debug output enabled
DEBUG 10:50:00.148893 config.go:119: DAOS Client config read from /home/tanabarr/projects/daos_m/utils/config/daos.yml
localhost:10001: connected
Hosts           SCM Total               NVMe Total
-----           ---------               ----------
localhost       5.79TB (2 namespaces)   3.03TB (3 controllers)
```

File: src/control/client/config_test.go [hunk 1]
```diff
@@ -63,33 +78,32 @@ func TestLoadConfigDefaultsNoFile(t *testing.T) {
 	log, buf := logging.NewTestLogger(t.Name())
 	defer common.ShowBufferOnFailure(t, buf)
 
-	defaultConfig := getDefaultConfig(t)
+	defaultConfig, f, closure := getDefaultConfig(t, log)
+	f.Close()
+	defer closure()
```
Comment (mjmac) [inline src/control/client/config_test.go]: nit: Naming it closure() isn't helpful for the reader. Yes, it is a closure, but what does it do? A better name/convention would be to call it cleanup() or otherwise allow the reader to understand what it's for without having to go look at the source of getDefaultConfig()

File: src/control/common/file_utils.go [hunk 2]
```diff
@@ -238,26 +229,53 @@ func Run(cmd string) error {
...
+		return inPath, nil
+	}
+
+	selfPath, err := os.Readlink("/proc/self/exe")
+	if err != nil {
+		return "", errors.Wrap(err, "unable to determine path to self")
+	}
+
+	return path.Join(path.Dir(selfPath), inPath), nil
+}
```
Comment (mjmac) [inline src/control/common/file_utils.go]: This doesn't sit right with me. As a control plane developer using our internal APIs, I would have to go read the source of these helper functions to discover that there is a potential plot twist: If the supplied path is an absolute path, then it's returned unmodified. That behavior doesn't match the function names or descriptions.

I would prefer to move the responsibility for checking path.IsAbs() and making a decision about what happens next to the caller. It might be slightly more code, but it's far more readable and easier to reason about.

If you want to make config-specific helpers that encapsulate this logic that would be fine, too. The helpers here in common are intended to be general-purpose and so their behavior should be straightforward and generally applicable.

File: src/control/security/config.go [hunk 3]
```diff
@@ -49,6 +50,10 @@ type TransportConfig struct {
 	CertificateConfig `yaml:",inline"`
 }
 
+func (tc *TransportConfig) String() string {
+	return fmt.Sprintf("allow insecure: %v", tc.AllowInsecure)
+}
+
```
Comment (mjmac) [inline src/control/security/config.go]: This seems unrelated.

Comment (tanabarr) [reply src/control/security/config.go]: this is so we can diff a string representation of a struct with transport config pointers
